prow/cmd/peribolos/main.go

```diff
@@ -870,3 +995,335 @@ func configureTeamMembers(client teamMembersClient, id int, team org.Team, invit
+		}
+		for _, u := range outsideCollaborators {
+			users = append(users, u.Login)
+		}
+	} else {
+		users = members.List()
+	}
+
+	for _, user := range users {
+		permission, err := client.GetUserPermission(orgName, repo, user)
+		if err != nil {
+			return err
+		}
+		switch permission {
+		case string(org.Admin):
+			haveAdmin.Insert(user)
+		case string(org.Write):
+			haveWrite.Insert(user)
+		case string(org.Read):
+			// don't consider read for members for now since all of them have implicit read access
+			if isOutsideCollaborator {
+				haveRead.Insert(user)
+			}
+		}
+	}
+
+	have := permissionLevels{admin: haveAdmin, write: haveWrite, read: haveRead}
+	want := permissionLevels{admin: wantAdmin, write: wantWrite, read: wantRead}
+	have.normalize()
+	want.normalize()
+
+	// figure out who to remove
+	remove := have.all().Difference(want.all())
+
+	// sanity check changes
+	if d := float64(len(remove)) / float64(len(have.all())); d > opt.maximumDelta {
+		return fmt.Errorf("cannot delete %d repo collaborators or %.3f of %s/%s (exceeds limit of %.3f)", len(remove), d, orgName, repo, opt.maximumDelta)
+	}
+
+	adder := func(user, permission string) error {
+		if invitees.Has(user) { // Do not add them, as this causes another invite.
+			logrus.Infof("Waiting for %s to accept invitation as a collaborator to %s/%s", user, orgName, repo)
+			return nil
+		}
+
+		isMember := members.Has(user)
+		if isMember && isOutsideCollaborator {
+			logrus.Infof("Cannot add/update %s as outside collaborator to %s/%s since they are a member", user, orgName, repo)
+			return nil
+		}
+
+		if !isMember && !isOutsideCollaborator {
+			logrus.Infof("%s is not a member of the org %s, retry as an outside collaborator or apply for membership", user, orgName)
+			return nil
+		}
+
+		err := client.UpdateCollaborator(orgName, repo, user, permission)
+		if err != nil {
+			logrus.WithError(err).Warnf("UpdateCollaborator(%s, %s, %s, %s) failed", orgName, repo, user, permission)
+		} else {
+			logrus.Infof("Set %s with %s access to %s/%s", user, permission, orgName, repo)
+		}
+		return err
+	}
+
+	remover := func(user string) error {
+		err := client.RemoveCollaborator(orgName, repo, user)
+		if err != nil {
+			logrus.WithError(err).Warnf("RemoveCollaborator(%s, %s, %s) failed", orgName, repo, user)
+		}
+		return err
+	}
+
+	return configureCollaborators(have, want, invitees, adder, remover)
+}
+
+func configureTeamCollaborators(opt options, client repoClient, orgName, repo string, config org.UserPermissionLevel, invitees sets.String) error {
+	// get desired state
+	wantAdmin := sets.NewString(config.Admin...)
+	wantWrite := sets.NewString(config.Write...)
+	wantRead := sets.NewString(config.Read...)
+
+	// get current state
+	teams, err := client.ListTeamsForRepo(orgName, repo)
+	if err != nil {
+		return err
+	}
+	var haveAdmin, haveWrite, haveRead sets.String
+	namesToID := map[string]int{}
+	for _, team := range teams {
+		namesToID[team.Name] = team.ID
```

Comment:
I'm not entirely sure if this is safe to do?

My motivation for doing this was to avoid storing the team metadata in `Repos`, since that leads to the source of truth about the team in two places - in the top-level `Teams` + here. Also, it's easier to understand on the basis of names, instead of IDs.

Reply:
Agree that peribolos should require uniquely named team names.

Can we do something to ensure that we always choose the same ID for a given name? See https://github.com/kubernetes/test-infra/blob/5475440d76f9039f7e1a5fa86c2f85ea8414b093/prow/cmd/peribolos/main.go#L512-L526